cli/command_restore.go

```diff
@@ -0,0 +1,81 @@
+package cli
+
+import (
+	"context"
+	"io"
+	"os"
+	"path"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/object"
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+)
+
+var (
+	restoreSnapshotID = restoreCommands.Arg("snapshot-id", "Snapshot ID from which to restore.").Required().String()
+	restoreTargetPath = restoreCommands.Arg("target-path", "Path to which the snapshot must be restored.").Required().String()
+)
+
+func runRestoreCommand(ctx context.Context, rep *repo.Repository) error {
+	oid, err := parseObjectID(ctx, rep, *restoreSnapshotID)
+	if err != nil {
+		return err
+	}
+	targetPath := *restoreTargetPath
+
+	return restoreRecursively(ctx, rep, targetPath, oid)
+}
+
+func init() {
+	restoreCommands.Action(repositoryAction(runRestoreCommand))
+}
+
+func restoreRecursively(ctx context.Context, rep *repo.Repository, targetPath string, oid object.ID) error {
+	d := snapshotfs.DirectoryEntry(rep, oid, nil)
+
+	entries, err := d.Readdir(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range entries {
+		restorePath := path.Join(targetPath, e.Name())
+		objectID := e.(object.HasObjectID).ObjectID()
+		// Restore directories recursively
+		if e.Mode().IsDir() {
+			// Create the directory
+			if direrr := os.MkdirAll(restorePath, 0777); direrr != nil {
```

Comment:
Could the initial permission be `0700` ?

Reply:
What happens when restoring to a directory that already exists? How should that be handled?

Reply:
1. Yes, we can use `0700`
2. `os.MkdirAll` will handle an already existing directory. Returns `nil` if already exists.
```
 If path is already a directory, MkdirAll does nothing and returns nil.
```

Reply:
Then, probably a stat afterwards is needed to ensure the directory can be written (either as owner, as group or other) . If it cannot be written, but it's owned by the user that the tool is running as, then perform the "chmod"

Reply:
> 2. `os.MkdirAll` will handle an already existing directory. Returns `nil` if already exists.

The questions about an existing directory is really about what happens when the directory already exists and it cannot be written to.

Also somewhat related:
* what happens if there is an entry with the same name, but it is not a directory?
* what happens to other files already present in a directory that are not in the snapshot directory.

The implementation should err on the safe side by not overriding or deleting existing files.
That behavior should be documented, probably in the command documentation.